### ISSUE TITLE
Wpe-2.38 Deadlock in WebKitMediaSrc (PrimeVideo)

### ISSUE THREAD

**Issue by asurdej-comcast** [Metrological/WebKitForWayland#1135]
There is a deadlock possible inside WebKit media src (MSE) between streaming thread handling CAPS event and pipeline flush in the main thread. This happens in case where there is a transform element in the pipeline that does transition NOT in place. Basetransform elem expects that it will allocate buffers so on CAPS change it triggers allocation negotiations (ALLOCATION query). In such case CAPS event becomes fully synchoronous as basetransform does ALLOCATOIN querry that is synchronous (serialized with data) and may block streaming thread. If the pipeline is paused and sink thread doesn't accept any data this will block CAPS event until pipeline is unpaused or flushed. But flush requires a lock that streaming thread is holding (DataMutexLocker streamingMembers { stream->streamingMembersDataMutex };)

Unfortunately I don't have good test case for this yet and this won't happen in all devices as some of them don't have basetransform not in place elements. It is visible with PrimeVideo app because video stream provides new caps (codec_data is slightly different) every like 1 sec so it is very easy to reproduce. But potentially this can happen on every CAPS change, like e.g. resolution change but it doesn't happen very frequently.

I think that the assumption that CAPS event is non-blocking is not fully correct in some cases. To fix this we need to either push CAPS without lock or make flush to not require the same lock.

Interesting parts:
1) 
```
webKitMediaSrcLoop():
...
    DataMutexLocker streamingMembers { stream->streamingMembersDataMutex };
...
        if (!gst_caps_is_equal(gst_sample_get_caps(sample.get()), streamingMembers->previousCaps.get())) {
            // This sample needs new caps (typically because of a quality change).
            GST_DEBUG_OBJECT(pad, "Pushing new CAPS event: %" GST_PTR_FORMAT, gst_sample_get_caps(sample.get()));
            bool result = gst_pad_push_event(stream->pad.get(), gst_event_new_caps(gst_sample_get_caps(sample.get())));
            GST_DEBUG_OBJECT(pad, "CAPS event pushed, result = %s.", boolForPrinting(result));
            ASSERT(result);
            streamingMembers->previousCaps = gst_sample_get_caps(sample.get());
        }
```
2)
```
static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
{
...
    {
        DataMutexLocker streamingMembers { stream->streamingMembersDataMutex };
        if (!streamingMembers->hasPoppedFirstObject) {
...
```

Stacktraces:
1) Video streaming thread is waiting for a sink to handle ALLOCATION query triggered from basetransform CAPS change
```
Breakpoint 1 at 0xb1745240: file ../sysdeps/unix/sysv/linux/arm/syscall.S, line 37.
(gdb) bt
#0  syscall () at ../sysdeps/unix/sysv/linux/arm/syscall.S:37
#1  0xb0628714 in g_cond_wait (cond=cond@entry=0xa6003d94, mutex=mutex@entry=0x7e4fe290) at ../glib-2.62.4/glib/gthread-posix.c:1422
#2  0xa2daaa38 in gst_multi_queue_sink_query (pad=<optimized out>, parent=0x7e4fe170, query=0xaa213c90) at ../gstreamer-1.18.5/plugins/elements/gstmultiqueue.c:2789
#3  0xb04d6d90 in gst_pad_query (pad=pad@entry=0xa6013060, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#4  0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0xa3d142b8, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#5  0xb04d7644 in query_forward_func (pad=0xa3d142b8, data=0x850fb570) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#6  0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e53e778, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fb570) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#7  0xb04d603c in gst_pad_query_default (pad=pad@entry=0x7e53e778, parent=parent@entry=0xa3d142b8, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#8  0xa375e5c6 in gst_parse_pad_query (pad=0x7e53e778, parent=0xa3d142b8, query=0xaa213c90) at ../gst-plugins-base-1.18.5/gst/playback/gstparsebin.c:4185
#9  0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e53e778, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#10 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0xa6034618, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#11 0xb04d7644 in query_forward_func (pad=0xa6034618, data=0x850fb808) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#12 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e51bda0, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fb808) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#13 0xb04d603c in gst_pad_query_default (pad=pad@entry=0x7e51bda0, parent=parent@entry=0xa6026308, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#14 0xafae25d6 in gst_base_parse_sink_query_default (parse=0xa6026308, query=0xaa213c90) at ../gstreamer-1.18.5/libs/gst/base/gstbaseparse.c:1613
#15 0xafae038a in gst_base_parse_sink_query (pad=<optimized out>, parent=0xa6026308, query=0xaa213c90) at ../gstreamer-1.18.5/libs/gst/base/gstbaseparse.c:1634
#16 0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e51bda0, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#17 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0xaa27d0f8, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#18 0xb04d7644 in query_forward_func (pad=0xaa27d0f8, data=0x850fbac0) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#19 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0xaa27d250, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fbac0) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#20 0xb04d603c in gst_pad_query_default (pad=0xaa27d250, parent=<optimized out>, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#21 0xb04d6d90 in gst_pad_query (pad=pad@entry=0xaa27d250, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#22 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0x7e53ee80, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#23 0xb04d7644 in query_forward_func (pad=0x7e53ee80, data=0x850fbd18) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#24 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e4eebb8, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fbd18) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#25 0xb04d603c in gst_pad_query_default (pad=0x7e4eebb8, parent=<optimized out>, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#26 0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e4eebb8, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#27 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0x7e53ed18, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#28 0xb04d7644 in query_forward_func (pad=0x7e53ed18, data=0x850fbf70) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#29 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e53d1b0, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fbf70) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#30 0xb04d603c in gst_pad_query_default (pad=0x7e53d1b0, parent=<optimized out>, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#31 0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e53d1b0, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#32 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0x7e53d490, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#33 0xb04d7644 in query_forward_func (pad=0x7e53d490, data=0x850fc1c8) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#34 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e4ef780, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fc1c8) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#35 0xb04d603c in gst_pad_query_default (pad=0x7e4ef780, parent=<optimized out>, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#36 0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e4ef780, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#37 0xb04d71f0 in gst_pad_peer_query (pad=pad@entry=0xaa27dd10, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#38 0xb04d7644 in query_forward_func (pad=0xaa27dd10, data=0x850fc420) at ../gstreamer-1.18.5/gst/gstpad.c:3453
#39 0xb04d5dfa in gst_pad_forward (pad=pad@entry=0x7e53a9f8, forward=0xb04d75d9 <query_forward_func>, user_data=user_data@entry=0x850fc420) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#40 0xb04d603c in gst_pad_query_default (pad=0x7e53a9f8, parent=<optimized out>, query=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:3524
#41 0xb04d6d90 in gst_pad_query (pad=pad@entry=0x7e53a9f8, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4149
#42 0xb04d71f0 in gst_pad_peer_query (pad=0xaa27d908, query=query@entry=0xaa213c90) at ../gstreamer-1.18.5/gst/gstpad.c:4281
#43 0xafaffe2a in gst_base_transform_do_bufferpool (outcaps=0xa3d11590, trans=0xa2c0c758) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:947
#44 gst_base_transform_setcaps (trans=trans@entry=0xa2c0c758, pad=<optimized out>, incaps=<optimized out>) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1350
#45 0xafb010e2 in gst_base_transform_sink_eventfunc (trans=0xa2c0c758, event=0x7e5ad948) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1945
#46 0xb04d1a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xaa27da60, event=event@entry=0x7e5ad948, type=<optimized out>, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#47 0xb04d1fce in gst_pad_push_event_unchecked (pad=pad@entry=0xaa27de68, event=0x7e5ad948, type=type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#48 0xb04d236a in push_sticky (pad=0xaa27de68, ev=0x850fc7a8, user_data=0x850fc7ec) at ../gstreamer-1.18.5/gst/gstevent.h:412
#49 0xb04d0692 in events_foreach (pad=pad@entry=0xaa27de68, func=0xb04d22e5 <push_sticky>, user_data=user_data@entry=0x850fc7ec) at ../gstreamer-1.18.5/gst/gstpad.c:608
#50 0xb04d2292 in check_sticky (pad=pad@entry=0xaa27de68, event=event@entry=0x7e5ad948) at ../gstreamer-1.18.5/gst/gstpad.c:4063
#51 0xb04d8252 in gst_pad_push_event (pad=pad@entry=0xaa27de68, event=0x7e5ad948) at ../gstreamer-1.18.5/gst/gstpad.c:5632
#52 0xafaffdbe in gst_pad_set_caps (caps=0xa3d116c8, pad=0xaa27de68) at ../gstreamer-1.18.5/gst/gstcompat.h:59
#53 gst_base_transform_setcaps (trans=trans@entry=0x7e538548, pad=<optimized out>, incaps=<optimized out>) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1345
#54 0xafb010e2 in gst_base_transform_sink_eventfunc (trans=0x7e538548, event=0x7e5ad9d8) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1945
#55 0xb2661d1c in sinkEventHandler () at ../git/Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:471
#56 0xb04d1a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x7e53a090, event=event@entry=0x7e5ad9d8, type=<optimized out>, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#57 0xb04d1fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa60055e0, event=0x7e5ad9d8, type=type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#58 0xb04d236a in push_sticky (pad=0xa60055e0, ev=0x850fcab0, user_data=0x850fcaf4) at ../gstreamer-1.18.5/gst/gstevent.h:412
#59 0xb04d0692 in events_foreach (pad=pad@entry=0xa60055e0, func=0xb04d22e5 <push_sticky>, user_data=user_data@entry=0x850fcaf4) at ../gstreamer-1.18.5/gst/gstpad.c:608
#60 0xb04d2292 in check_sticky (pad=pad@entry=0xa60055e0, event=event@entry=0x7e5ad9d8) at ../gstreamer-1.18.5/gst/gstpad.c:4063
#61 0xb04d8252 in gst_pad_push_event (pad=0xa60055e0, event=0x7e5ad9d8) at ../gstreamer-1.18.5/gst/gstpad.c:5632
--Type <RET> for more, q to quit, c to continue without paging--
#62 0xb2670208 in webKitMediaSrcLoop () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:756
#63 0xb04f884c in gst_task_func (task=0xaa2ba418) at ../gstreamer-1.18.5/gst/gsttask.c:384
#64 0xb061545c in g_thread_pool_thread_proxy (data=<optimized out>) at ../glib-2.62.4/glib/gthreadpool.c:308
#65 0xb0614f2e in g_thread_proxy (data=0x7ddb6a90) at ../glib-2.62.4/glib/gthread.c:805
#66 0xafdbc79e in start_thread (arg=0xf714cfdd) at pthread_create.c:477
#67 0xb17471ac in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from ./lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```
2) Unfortunatelly video sink is paused and doesnt handle any buffers so above thread is just waiting inside gst_push_event()
3) Main thread is trying to flush it and unblock the sink. But it needs to get the lock that video streaming occupies
```
(gdb) bt
#0  __libc_do_syscall () at libc-do-syscall.S:49
#1  0xafdc137e in futex_wait_cancelable (private=0, expected=0, futex_word=0xad6f104c) at ../sysdeps/nptl/futex-internal.h:183
#2  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0xad6f1008, cond=0xad6f1020) at pthread_cond_wait.c:508
#3  __pthread_cond_wait (cond=0xad6f1020, mutex=0xad6f1008) at pthread_cond_wait.c:638
#4  0xb25f03d0 in WTF::ThreadCondition::timedWait () at ../git/Source/WTF/wtf/posix/ThreadingPOSIX.cpp:613
#5  0xb25b537a in WTF::ParkingLot::parkConditionallyImpl(void const*, WTF::ScopedLambda<bool ()> const&, WTF::ScopedLambda<void ()> const&, WTF::TimeWithDynamicClockType const&) () at ../git/Source/WTF/wtf/ParkingLot.cpp:595
#6  0xb25ae20a in WTF::ParkingLot::parkConditionally<WTF::ParkingLot::compareAndPark<unsigned char, unsigned char>(WTF::Atomic<unsigned char> const*, unsigned char)::{lambda()#1}, WTF::ParkingLot::compareAndPark<unsigned char, unsigned char>(WTF::Atomic<unsigned char> const*, unsigned char)::{lambda()#2}>(void const*, WTF::ParkingLot::compareAndPark<unsigned char, unsigned char>(WTF::Atomic<unsigned char> const*, unsigned char)::{lambda()#1} const&, WTF::ParkingLot::compareAndPark<unsigned char, unsigned char>(WTF::Atomic<unsigned char> const*, unsigned char)::{lambda()#2} const&, WTF::TimeWithDynamicClockType const&) () at ../git/Source/WTF/wtf/ParkingLot.h:82
#7  WTF::ParkingLot::compareAndPark<unsigned char, unsigned char> () at ../git/Source/WTF/wtf/ParkingLot.h:94
#8  WTF::LockAlgorithm<unsigned char, (unsigned char)1, (unsigned char)2, WTF::EmptyLockHooks<unsigned char> >::lockSlow () at ../git/Source/WTF/wtf/LockAlgorithmInlines.h:84
#9  0xb266e68a in WTF::Lock::lock () at WTF/Headers/wtf/Lock.h:66
#10 WTF::DataMutexLocker<Stream::StreamingMembers>::lock () at WTF/Headers/wtf/DataMutex.h:126
#11 WTF::DataMutexLocker<Stream::StreamingMembers>::DataMutexLocker () at WTF/Headers/wtf/DataMutex.h:71
#12 webKitMediaSrcStreamFlush () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:808
#13 0xb266eb76 in webKitMediaSrcSeek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:918
#14 webKitMediaSrcSendEvent () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:982
#15 webKitMediaSrcSendEvent () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:968
#16 0xb04bd472 in gst_element_send_event (element=0x7e2a6430, event=0xaa269390) at ../gstreamer-1.18.5/gst/gstelement.c:1947
#17 0xb266945e in WebCore::MediaPlayerPrivateGStreamerMSE::doSeek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:272
#18 0xb26681ea in WebCore::MediaPlayerPrivateGStreamerMSE::seek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:243
#19 0xb30a995e in WebCore::HTMLMediaElement::seekTask () at ../git/Source/WebCore/html/HTMLMediaElement.cpp:3322
#20 0xb2f6b6e2 in WTF::Function<void ()>::operator()() const () at WTF/Headers/wtf/Function.h:82
#21 WebCore::EventLoopFunctionDispatchTask::execute () at ../git/Source/WebCore/dom/EventLoop.cpp:160
#22 WebCore::EventLoop::run () at ../git/Source/WebCore/dom/EventLoop.cpp:124
#23 0xb2faeeda in WebCore::WindowEventLoop::didReachTimeToRun () at ../git/Source/WebCore/dom/WindowEventLoop.cpp:121
#24 0xb3326be4 in WebCore::ThreadTimers::sharedTimerFiredInternal () at ../git/Source/WebCore/platform/ThreadTimers.cpp:127
#25 0xb25edf2a in operator() () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:177
#26 _FUN () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:181
#27 0xb25ee4f8 in operator() () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:53
#28 _FUN () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:56
#29 0xb05fe864 in g_main_dispatch (context=0x7dd250b8) at ../glib-2.62.4/glib/gmain.c:3216
#30 g_main_context_dispatch (context=context@entry=0x7dd250b8) at ../glib-2.62.4/glib/gmain.c:3885
#31 0xb05fe9ce in g_main_context_iterate (context=0x7dd250b8, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib-2.62.4/glib/gmain.c:3958
#32 0xb05fed08 in g_main_loop_run (loop=0x7dd2fb78) at ../glib-2.62.4/glib/gmain.c:4152
#33 0xb25ee798 in WTF::RunLoop::run () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:108
#34 0xb1d90e1a in WebKit::AuxiliaryProcessMainBase<WebKit::WebProcess, true>::run () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:71
#35 WebKit::AuxiliaryProcessMainBase<WebKit::WebProcess, true>::run () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:58
#36 WebKit::AuxiliaryProcessMain<WebKit::WebProcessMainWPE> () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:97
#37 0xb16dd98e in __libc_start_main (main=0x7d628545, argc=3, argv=0xbbc34f74, init=<optimized out>, fini=0x7d628699, rtld_fini=0xb3f49099 <_dl_fini>, stack_end=0xbbc34f74) at libc-start.c:308
#38 0x7d62857c in ?? ()
```

**Comment by modeveci** [comment]
@asurdej-comcast I assume it is for wpe.2-38, if not please state which version.

**Comment by asurdej-comcast** [comment]
Yes, this is 2.38 only. Works fine with 2.28 (2.28 uses appsrc element that doesn't have this issue )

**Comment by eocanha** [comment]
Thank you so much for all the analysis. It's been pretty useful.

I've written a tentative solution that releases the lock before sending the caps event and reevaluates the flushing condition after reacquiring it (pausing the task and exiting webKitMediaSrcLoop() if a flush has happened while the thread wasn't in posession of the lock).

I've submitted the issue and patch upstream as https://bugs.webkit.org/show_bug.cgi?id=260067 and https://github.com/WebKit/WebKit/pull/16603.

While it's being reviewed, could you try to apply to your WPE code and check if it solves the deadlock for you without unintended consequences?

**Comment by asurdej-comcast** [comment]
@eocanha 
Thanks for the patch, looks good with prime video app, the issue is no longer there. I will schedule more testing with our QA to make sure no other app is broken.
However different hang seen, not sure if related but maybe you recognize it. Happened on initial seek while sending FLUSH_STOP for video. One time issue.
```
Thread 1 (Thread 0xae633010 (LWP 29595)):
#0  syscall () at ../sysdeps/unix/sysv/linux/arm/syscall.S:38
#1  0xb061c714 in g_cond_wait (cond=cond@entry=0x9a925a84, mutex=mutex@entry=0x9a925a34) at ../glib-2.62.4/glib/gthread-posix.c:1422
#2  0xb04c3800 in do_probe_callbacks (pad=pad@entry=0x9a925a28, info=info@entry=0xbbd16878, defaultval=defaultval@entry=GST_FLOW_OK) at ../gstreamer-1.18.5/gst/gstpad.c:3851
#3  0xb04c5d9a in gst_pad_push_event_unchecked (pad=pad@entry=0x9a925a28, event=0xa2b0b830, type=type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5448
#4  0xb04c636a in push_sticky (pad=0x9a925a28, ev=0xbbd16910, user_data=0xbbd16954) at ../gstreamer-1.18.5/gst/gstevent.h:412
#5  0xb04c4692 in events_foreach (pad=pad@entry=0x9a925a28, func=0xb04c62e5 <push_sticky>, user_data=user_data@entry=0xbbd16954) at ../gstreamer-1.18.5/gst/gstpad.c:608
#6  0xb04c6292 in check_sticky (pad=pad@entry=0x9a925a28, event=event@entry=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:4063
#7  0xb04cc252 in gst_pad_push_event (pad=pad@entry=0x9a925a28, event=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:5632
#8  0xb04cc3ea in event_forward_func (pad=0x9a925a28, data=0xbbd16a64) at ../gstreamer-1.18.5/gst/gstevent.h:412
#9  0xb04c9dfa in gst_pad_forward (pad=pad@entry=0x7f7e68e0, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd16a64) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#10 0xb04c9f4e in gst_pad_event_default (pad=0x7f7e68e0, parent=<optimized out>, event=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#11 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x7f7e68e0, event=event@entry=0xa2b0b830, type=<optimized out>, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#12 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa970dbf0, event=0xa2b0b830, type=type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#13 0xb04c636a in push_sticky (pad=0xa970dbf0, ev=0xbbd16bd8, user_data=0xbbd16c1c) at ../gstreamer-1.18.5/gst/gstevent.h:412
#14 0xb04c4692 in events_foreach (pad=pad@entry=0xa970dbf0, func=0xb04c62e5 <push_sticky>, user_data=user_data@entry=0xbbd16c1c) at ../gstreamer-1.18.5/gst/gstpad.c:608
#15 0xb04c6292 in check_sticky (pad=pad@entry=0xa970dbf0, event=event@entry=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:4063
#16 0xb04cc252 in gst_pad_push_event (pad=pad@entry=0xa970dbf0, event=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:5632
#17 0xb04cc3ea in event_forward_func (pad=0xa970dbf0, data=0xbbd16d2c) at ../gstreamer-1.18.5/gst/gstevent.h:412
#18 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0xa5f0c4b8, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd16d2c) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#19 0xb04c9f4e in gst_pad_event_default (pad=pad@entry=0xa5f0c4b8, parent=parent@entry=0x7f1dabf0, event=0xa2b0b830) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#20 0x9fe2f0dc in gst_concat_sink_event (pad=0xa5f0c4b8, parent=0x7f1dabf0, event=<optimized out>) at ../gstreamer-1.18.5/plugins/elements/gstconcat.c:704
#21 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa5f0c4b8, event=event@entry=0xa2b0b830, type=<optimized out>, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#22 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa8989768, event=0xa2b0b830, type=type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#23 0xb04c636a in push_sticky (pad=0xa8989768, ev=0xbbd16f98, user_data=0xbbd16fdc) at ../gstreamer-1.18.5/gst/gstevent.h:412
#24 0xb04c4692 in events_foreach (pad=pad@entry=0xa8989768, func=0xb04c62e5 <push_sticky>, user_data=user_data@entry=0xbbd16fdc) at ../gstreamer-1.18.5/gst/gstpad.c:608
#25 0xb04c6292 in check_sticky (pad=pad@entry=0xa8989768, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:4063
#26 0xb04cc252 in gst_pad_push_event (pad=pad@entry=0xa8989768, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5632
#27 0xb04cc3ea in event_forward_func (pad=0xa8989768, data=0xbbd170ec) at ../gstreamer-1.18.5/gst/gstevent.h:412
#28 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0xa898ae60, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd170ec) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#29 0xb04c9f4e in gst_pad_event_default (pad=0xa898ae60, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
--Type <RET> for more, q to quit, c to continue without paging--
#30 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa898ae60, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#31 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa8989d28, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#32 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0xa8989d28, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#33 0xb04cc3ea in event_forward_func (pad=0xa8989d28, data=0xbbd17314) at ../gstreamer-1.18.5/gst/gstevent.h:412
#34 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0x9a903bb0, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd17314) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#35 0xb04c9f4e in gst_pad_event_default (pad=0x9a903bb0, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#36 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x9a903bb0, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#37 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xaa175648, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#38 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0xaa175648, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#39 0x9fe44746 in gst_multi_queue_sink_event (pad=0xaa175388, parent=0x7f9b8408, event=0xa3c09900) at ../gstreamer-1.18.5/plugins/elements/gstmultiqueue.c:2605
#40 0xb04c5a0c in gst_pad_send_event_unchecked (pad=pad@entry=0xaa175388, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5856
#41 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0x7f9b6500, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#42 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0x7f9b6500, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#43 0xb04cc3ea in event_forward_func (pad=0x7f9b6500, data=0xbbd176f4) at ../gstreamer-1.18.5/gst/gstevent.h:412
#44 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0x9a9034a8, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd176f4) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#45 0xb04c9f4e in gst_pad_event_default (pad=0x9a9034a8, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#46 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x9a9034a8, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#47 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2bac720, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#48 0xb04cc268 in gst_pad_push_event (pad=0xa2bac720, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#49 0xafaf50b8 in gst_base_transform_sink_eventfunc (trans=0xa2bafd58, event=0xa3c09900) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1966
#50 0x9fe2cbec in gst_capsfilter_sink_event (trans=0xa2bafd58, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstevent.h:412
#51 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa2bacc80, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#52 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0x7fb8e068, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#53 0xb04cc268 in gst_pad_push_event (pad=0x7fb8e068, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#54 0xafadd18a in gst_base_parse_sink_event_default (parse=0xa3c48410, event=0xa3c09900) at ../gstreamer-1.18.5/libs/gst/base/gstbaseparse.c:1552
#55 0x9ca50b9e in gst_h265_sec_parse_event (parse=0xa3c48410, event=0xa3c09900) at ../../../git/gst-aml-drm-plugins-1.0/src/secure_parse/gsth265_sec_parse.c:2253
#56 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x7faf63c8, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#57 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2bb99d8, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#58 0xb04cc268 in gst_pad_push_event (pad=0xa2bb99d8, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#59 0x9fe59cd2 in gst_type_find_element_sink_event (pad=<optimized out>, parent=0x7fa81b18, event=0xa3c09900) at ../gstreamer-1.18.5/plugins/elements/gsttypefindelement.c:746
#60 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa2bac5c8, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#61 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2b06088, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#62 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0xa2b06088, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#63 0xb04cc3ea in event_forward_func (pad=0xa2b06088, data=0xbbd17f04) at ../gstreamer-1.18.5/gst/gstevent.h:412
#64 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0x9a920310, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd17f04) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#65 0xb04c9f4e in gst_pad_event_default (pad=0x9a920310, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#66 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x9a920310, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#67 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0x9a903070, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#68 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0x9a903070, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#69 0xb04cc3ea in event_forward_func (pad=0x9a903070, data=0xbbd1812c) at ../gstreamer-1.18.5/gst/gstevent.h:412
#70 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0xa8989318, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd1812c) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#71 0xb04c9f4e in gst_pad_event_default (pad=0xa8989318, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#72 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa8989318, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#73 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0x7f7e61b0, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#74 0xb04cc268 in gst_pad_push_event (pad=pad@entry=0x7f7e61b0, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#75 0xb04cc3ea in event_forward_func (pad=0x7f7e61b0, data=0xbbd18354) at ../gstreamer-1.18.5/gst/gstevent.h:412
#76 0xb04c9dfa in gst_pad_forward (pad=pad@entry=0xa898ab90, forward=0xb04cc379 <event_forward_func>, user_data=user_data@entry=0xbbd18354) at ../gstreamer-1.18.5/gst/gstpad.c:3079
#77 0xb04c9f4e in gst_pad_event_default (pad=0xa898ab90, parent=<optimized out>, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:3176
#78 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa898ab90, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#79 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2bac068, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#80 0xb04cc268 in gst_pad_push_event (pad=0xa2bac068, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#81 0x9fe59cd2 in gst_type_find_element_sink_event (pad=<optimized out>, parent=0xaa1750a0, event=0xa3c09900) at ../gstreamer-1.18.5/plugins/elements/gsttypefindelement.c:746
#82 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa2b69350, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#83 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2b691f8, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#84 0xb04cc268 in gst_pad_push_event (pad=0xa2b691f8, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#85 0xafaf50b8 in gst_base_transform_sink_eventfunc (trans=0xa3c12170, event=0xa3c09900) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1966
#86 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0x7f7e58b8, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
#87 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0xa2b69e10, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#88 0xb04cc268 in gst_pad_push_event (pad=0xa2b69e10, event=event@entry=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#89 0xafaf50b8 in gst_base_transform_sink_eventfunc (trans=0xa2b050b0, event=0xa3c09900) at ../gstreamer-1.18.5/libs/gst/base/gstbasetransform.c:1966
#90 0xb265cf74 in sinkEventHandler () at ../git/Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:471
#91 0xb04c5a9a in gst_pad_send_event_unchecked (pad=pad@entry=0xa2b69cb8, event=event@entry=0xa3c09900, type=<optimized out>, type@entry=320) at ../gstreamer-1.18.5/gst/gstpad.c:5857
--Type <RET> for more, q to quit, c to continue without paging--
#92 0xb04c5fce in gst_pad_push_event_unchecked (pad=pad@entry=0x9a903778, event=event@entry=0xa3c09900, type=320, type@entry=GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) at ../gstreamer-1.18.5/gst/gstpad.c:5501
#93 0xb04cc268 in gst_pad_push_event (pad=0x9a903778, event=0xa3c09900) at ../gstreamer-1.18.5/gst/gstpad.c:5638
#94 0xb26693cc in webKitMediaSrcStreamFlush () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:900
#95 0xb2669be6 in webKitMediaSrcSeek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:928
#96 webKitMediaSrcSendEvent () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:992
#97 webKitMediaSrcSendEvent () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:978
#98 0xb04b1472 in gst_element_send_event (element=0x7f7e2608, event=0xa2b0bb00) at ../gstreamer-1.18.5/gst/gstelement.c:1947
#99 0xb26643f6 in WebCore::MediaPlayerPrivateGStreamerMSE::doSeek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:265
#100 0xb2663182 in WebCore::MediaPlayerPrivateGStreamerMSE::seek () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:236
#101 0xb30a473e in WebCore::HTMLMediaElement::seekTask () at ../git/Source/WebCore/html/HTMLMediaElement.cpp:3322
#102 0xb2f664c2 in WTF::Function<void ()>::operator()() const () at WTF/Headers/wtf/Function.h:82
#103 WebCore::EventLoopFunctionDispatchTask::execute () at ../git/Source/WebCore/dom/EventLoop.cpp:160
#104 WebCore::EventLoop::run () at ../git/Source/WebCore/dom/EventLoop.cpp:124
#105 0xb2fa9cba in WebCore::WindowEventLoop::didReachTimeToRun () at ../git/Source/WebCore/dom/WindowEventLoop.cpp:121
#106 0xb3321f74 in WebCore::ThreadTimers::sharedTimerFiredInternal () at ../git/Source/WebCore/platform/ThreadTimers.cpp:127
#107 0xb25e9e4a in operator() () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:177
#108 _FUN () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:181
#109 0xb25ea418 in operator() () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:53
#110 _FUN () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:56
#111 0xb05f2864 in g_main_dispatch (context=0x7f1ca598) at ../glib-2.62.4/glib/gmain.c:3216
#112 g_main_context_dispatch (context=context@entry=0x7f1ca598) at ../glib-2.62.4/glib/gmain.c:3885
#113 0xb05f29ce in g_main_context_iterate (context=0x7f1ca598, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib-2.62.4/glib/gmain.c:3958
#114 0xb05f2d08 in g_main_loop_run (loop=0x7f1d5800) at ../glib-2.62.4/glib/gmain.c:4152
#115 0xb25ea6b8 in WTF::RunLoop::run () at ../git/Source/WTF/wtf/glib/RunLoopGLib.cpp:108
#116 0xb1d8cce2 in WebKit::AuxiliaryProcessMainBase<WebKit::WebProcess, true>::run () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:71
#117 WebKit::AuxiliaryProcessMainBase<WebKit::WebProcess, true>::run () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:58
#118 WebKit::AuxiliaryProcessMain<WebKit::WebProcessMainWPE> () at ../git/Source/WebKit/Shared/AuxiliaryProcessMain.h:97
#119 0xb16d998e in __libc_start_main (main=0x7d650545, argc=3, argv=0xbbd18f44, init=<optimized out>, fini=0x7d650699, rtld_fini=0xb3f44099 <_dl_fini>, stack_end=0xbbd18f44) at libc-start.c:308
#120 0x7d65057c in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```
Streaming thread waiting for buffer or flush. I think this is video thread as audio was flushed before so thread should be stopped
```
Thread 26 (Thread 0x869fe260 (LWP 2370)):
#0  __libc_do_syscall () at libc-do-syscall.S:48
#1  0xafdb537e in futex_wait_cancelable (private=0, expected=0, futex_word=0x70de2048) at ../sysdeps/nptl/futex-internal.h:183
#2  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x70de2008, cond=0x70de2020) at pthread_cond_wait.c:508
#3  __pthread_cond_wait (cond=0x70de2020, mutex=0x70de2008) at pthread_cond_wait.c:638
#4  0xb25ec2f0 in WTF::ThreadCondition::timedWait () at ../git/Source/WTF/wtf/posix/ThreadingPOSIX.cpp:613
#5  0xb25b152a in WTF::ParkingLot::parkConditionallyImpl(void const*, WTF::ScopedLambda<bool ()> const&, WTF::ScopedLambda<void ()> const&, WTF::TimeWithDynamicClockType const&) () at ../git/Source/WTF/wtf/ParkingLot.cpp:595
#6  0xb266b670 in WTF::ParkingLot::parkConditionally<WTF::Condition::waitUntilUnchecked<WTF::Lock>(WTF::Lock&, WTF::TimeWithDynamicClockType const&)::{lambda()#1}, WTF::Condition::waitUntilUnchecked<WTF::Lock>(WTF::Lock&, WTF::TimeWithDynamicClockType const&)::{lambda()#2}>(void const*, WTF::Condition::waitUntilUnchecked<WTF::Lock>(WTF::Lock&, WTF::TimeWithDynamicClockType const&)::{lambda()#1} const&, WTF::Condition::waitUntilUnchecked<WTF::Lock>(WTF::Lock&, WTF::TimeWithDynamicClockType const&)::{lambda()#2} const&, WTF::TimeWithDynamicClockType const&) () at WTF/Headers/wtf/ParkingLot.h:82
#7  WTF::Condition::waitUntilUnchecked<WTF::Lock> () at WTF/Headers/wtf/Condition.h:192
#8  WTF::Condition::waitUntil () at WTF/Headers/wtf/Condition.h:77
#9  WTF::Condition::wait () at WTF/Headers/wtf/Condition.h:127
#10 wait<webKitMediaSrcLoop(void*)::<lambda()> > () at WTF/Headers/wtf/Condition.h:141
#11 webKitMediaSrcLoop () at ../git/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:729
#12 0xb04ec84c in gst_task_func (task=0xaa1524c0) at ../gstreamer-1.18.5/gst/gsttask.c:384
#13 0xb060945c in g_thread_pool_thread_proxy (data=<optimized out>) at ../glib-2.62.4/glib/gthreadpool.c:308
#14 0xb0608f2e in g_thread_proxy (data=0xa9705b20) at ../glib-2.62.4/glib/gthread.c:805
#15 0xafdb079e in start_thread (arg=0xe2c4c5bc) at pthread_create.c:477
#16 0xb17431ac in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /home/asurdej/cores/HSTP12MWR_VBN_23Q3_sprint_20230822080625sdy_wpe-2.38_prime/lib/libc.so.6
```
[RDK-43710_prime_initial_seek_crash_bt_full.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/12417503/RDK-43710_prime_initial_seek_crash_bt_full.txt)

**Comment by eocanha** [comment]
> I've written a tentative solution that releases the lock before sending the caps event and reevaluates the flushing condition after reacquiring it (pausing the task and exiting webKitMediaSrcLoop() if a flush has happened while the thread wasn't in posession of the lock).
> I've submitted the issue and patch upstream as https://bugs.webkit.org/show_bug.cgi?id=260067 and [WebKit/WebKit#16603](https://github.com/WebKit/WebKit/pull/16603).

The patch has landed upstream as https://commits.webkit.org/267276@main and on wpe-2.38 as https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/ea795729d0adc255a2e1e6d28fcb57eb1eba2792

I haven't had time yet to analyze the new deadlock reported.

**Comment by asurdej-comcast** [comment]
The fix was included in wpe-2.38 branch and it fixes the issue. We haven't spot any regression so far.
Closing this ticket as original bug is resolved.

I will monitor new deadlock and fill another ticket if neccessary.
@eocanha Thanks for you help
